net: lwm2m: Add LWM2M object 9 software management

MrOneTwo opened the pull request:
Adds support for LWM2M object 9 Software management.
This is implemented according to this release:
http://openmobilealliance.org/release/LWM2M_SWMGMT/V1_0_1-20200616-A/

Note that the XML is lacking some resources and for that reason those
resources are not included. This is a known problem by OMA and will be
fixed in a later releases.

This uses the lwm2m_pull_context to pull binaries in case
FIRMWARE_PULL_SUPPORT is enabled

This PR has been introduced before by @DunderRoffe but he closed it because he couldn't finalize it (lack of time). I'm trying to pick up the torch. There has been a few small bugs fixed between then and this PR.

MrOneTwo committed:
net: lwm2m: Make lwm2m_pull_context reusable

Add a semaphore to control that no collisions occur when multiple
sources want to use the pull_context

Add struct firmware_pull_context *ctx as an argument to the result_cb of
the context. This allows the receiver to do some kind of differentiation
on the source.

Signed-off-by: Michal Ciesielski <michal.m.ciesielski@voiapp.io>

MrOneTwo committed:
net: lwm2m: FW Pull separate generic pull logic

Make it possible to reuse the pull logic from firmware_pull.c by
separating it to a separate file.

The firmware_pull_context is still owned and statically allocated in
firmware_pull.c and is being passed into lwm2m_pull_context.c as a
pointer.

In other words, pull_context, does not keep any state except for a
pointer to the context currently in use.

Signed-off-by: Michal Ciesielski <michal.m.ciesielski@voiapp.io>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+	instance->pull_context.obj_inst_id = instance->obj_inst_id;
+
+	strncpy(instance->pull_context.uri, instance->package_uri, PACKAGE_URI_LEN - 1);
+	error_code = lwm2m_pull_context_start_transfer(&instance->pull_context, K_NO_WAIT);
```
Is the PR complete? I don't see this being defined anywhere.

MrOneTwo replied:
It's not. I've realized there are commits that this depends on and I'm working on integrating them into this PR. Sorry for the confusion.

rlubos commented:
It seems that we have an alternative proposal in https://github.com/zephyrproject-rtos/zephyr/pull/40628. 
Before we proceed, we need to settle which PR should take precedence, both need some work and there's no point addressing issues in both of them since only one can get in.

@MrOneTwo @jarlamsa What's your opinion on this? I'm asking specifically which one of you is more determined to put effort into addressing review comments and proceeding with the PR.

MrOneTwo commented:
@rlubos @jarlamsa I'll try to address the issues in this PR during the coming week. All I can say is that I'm pretty committed to getting this in soon.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -16,7 +16,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+static void set_update_result_from_error(struct firmware_pull_context *context,
+                                         int error_code)
 {
 	if (!error_code) {
```
Woudl be good to use `ARG_UNUSED()` on `context` here.

jarlamsa replied:
The function is now a bit misleading as setting update result from error when there is no error suggests something else.

MrOneTwo replied:
Hmmm @jarlamsa I understand but since it's a static function of this file do you think this is a problem? I think it's pretty readable that if no error is present then we report success. Maybe we can make the name more generic? `set_update_result`? I'm open to suggestions :)

@rlubos Added `ARG_UNUSED()`

jarlamsa replied:
`set_update_result` would be good name for this

MrOneTwo replied:
fixed

rlubos commented on subsys/net/lib/lwm2m/lwm2m_pull_context.c:
```diff
@@ -344,6 +359,18 @@ static void firmware_transfer(struct k_work *work)
+
+	delayed_work = CONTAINER_OF(work, struct k_delayed_work, work);
+
+	ret = k_sem_take(&lwm2m_pull_sem, K_NO_WAIT);
```
Isn't it a bit of a waste, to keep separate PULL context per each object instance, but then limit the usage to a single download at once (note that this structure is pretty large due to fields like `uri` or `lwm2m_ctx`)?  Perhaps it'd be more economic to have a single context, and then verify inside `lwm2m_pull_context_start_transfer()` if it's in use (it'd then be the caller's responsibility to try again later).

MrOneTwo replied:
Agree. I'll go single context.

MrOneTwo replied:
@rlubos I would appreciate you taking a look at this solution https://github.com/zephyrproject-rtos/zephyr/pull/40552/commits/afbb37219139acd2467602f543c850be9e6ed23e - this is handling failure when a package is already being pulled.

Safety check if the work is already scheduled might be unnecessary but I'd really like your input on this.

rlubos replied:
I just wonder, do we really need `firmware_work` in this case? This initial check to determine if download is already ongoing looks fine, but I don't see why would need a work here, `firmware_transfer()` could just be a regular function called from that context.

I'd rather see a delayed work use on the caller side - if the function returns `-EALREADY` it should set a delayed work and retry later.

I also don't think we should expect `struct firmware_pull_context *` as an argument for the `lwm2m_pull_context_start_transfer()` - since the only instance in use would exist in the `lwm2m_pull_context.c`, the callers don't even need to be aware of this structure. We only need to provide a few information needed to start the download - URI string pointer, instance ID, callback pointers, perhaps nicely wrapped into some helper struct. But this struct does not need to be in the static storage, it could exist on a stack, since all of the data would be copied into the pull context when the download is started.

MrOneTwo replied:
Hard to disagree with anything you've pointed out here.

I'll get to implementing these changes.

rlubos commented on include/net/lwm2m.h:
```diff
@@ -365,6 +366,99 @@ struct coap_block_context *lwm2m_firmware_get_block_context();
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_activate_cb(uint16_t obj_inst_id, lwm2m_engine_user_cb_t cb);
```
I think `lwm2m_engine_execute_cb_t` would be more appropriate here. Same for other actions (deactivate/install/uninstall).

MrOneTwo replied:
Fixed

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+#define SWMGMT_PASSWORD_ID 15
+#define SWMGMT_MAX_ID 16
+
+#define PACKAGE_NAME_LEN 255
+#define PACKAGE_VERSION_LEN 255
+#define PACKAGE_URI_LEN 255
```
It's a lot of space for the package name/version, needed per object instance. I think it'd be good to make those values configurable with Kconfig.

MrOneTwo replied:
done and I'm setting some lower default values

rlubos commented on include/net/lwm2m.h:
```diff
@@ -365,6 +366,99 @@ struct coap_block_context *lwm2m_firmware_get_block_context();
+ *
+ * @return 0 on success, otherwise a negative integer.
+ */
+int lwm2m_swmgmt_set_read_package_cb(uint16_t obj_inst_id, lwm2m_engine_get_data_cb_t cb);
```
I just wonder, is this really needed? There's no special action needed on the object implementation side before the callback is exectued, so the app could simply use generic `lwm2m_engine_register_read_callback()`.

MrOneTwo replied:
Hmmm... on a functional level I don't think this is needed - meaning this doesn't provide unique functionality compared to `lwm2m_engine_register_read_callback`. The thing is that *read package* callback reads the resource's version attribute. It's not a typical *read the data callback*. So technically the user can implement reading the version attribute and pass that via the `lwm2m_engine_register_read_callback` but I think that would be a bit misleading. Having a separate function creates a distinction between *set read data callback* and *set read package version attribute*.

I hope I'm getting this right ;)

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+static struct lwm2m_swmgmt_data *find_index(uint16_t obj_inst_id)
+{
+	int index;
+
```
Extra empty line. There are other places as well like this, in general we don't put more than one empty line between instructions in the code.

MrOneTwo replied:
fixed

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+static int lwm2m_swmgmt_init(const struct device *dev)
+{
+	swmgmt.obj_id = LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID;
+	swmgmt.fields = fields;
```
You should also set version_major/minor fields.

MrOneTwo replied:
Setting it to 1.1

rlubos replied:
Why 1.1? I see in [OMA registry](https://technical.openmobilealliance.org/OMNA/LwM2M/LwM2MRegistry.html) that this object is only available in version 1.0.

rlubos reviewed (commented):
Initial high level feedback. Still need to take a closer look at the SW managmament state machine, but it looks pretty good already.

rlubos commented:
@MrOneTwo Thanks again for the contribution, we've agreed with @jarlamsa to focus on this PR since it brings more functionality to Zephyr (i.e. PULL mode support).

jarlamsa commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+		handle_event(instance, EVENT_PKG_WRITTEN);
+
+		/* Do this somehow? */
+		handle_event(instance, EVENT_PKG_INTEGRITY_VERIFIED);
```
There should be some possibility to have an external function doing the package integrity verification and reporting the results to the object. Currently I can't see a clear path leading here.

MrOneTwo replied:
When you say `external` do you mean something settable by the Zephyr's user? A callback that they can set or should it be a static function in the lwm2m scope? Is it fine if we leave that for a future PR or do you think this has to be a part of this PR?

I'd obviously like it to do this but this PR is already pretty big. The comment needs fixing for sure if we choose to implement this later.

MrOneTwo replied:
I've implemented this but I did per download request. Do you see any value in storing a verify function per object instance?

jarlamsa commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,743 @@
+static void set_update_result_from_error(struct firmware_pull_context *context, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance;
+	instance = CONTAINER_OF(context, struct lwm2m_swmgmt_data, pull_context);
```
There isn't `firmware_pull_context` in `lwm2m_swmgmt_data` when `CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT` isn't enabled. Is this supposed only to be used with the PULL-method? What about PUSH-method?

MrOneTwo replied:
This doesn't support PUSH atm. This is something that should be implemented in the future. Any safety check that we could introduce here?

rlubos replied:
That's a bit unfortunate to be honest. Do  Iunderstand correctly, that the only limitation is the PULL context provided to the result callback? Perhaps instead of providing this pointer, we could simple feed the callback with the instance number? On the FW update object side it's not used anyway, and here we only use the pointer to obtain the acual object instance - we could achieve the same with instance number and `find_index()`

jarlamsa replied:
Lack of PUSH is indeed unfortunate. Any estimates that how long it would take to implement?

MrOneTwo replied:
Since I've inherited this PR I cannot answer how long it would take, right now. I'm just not knowledgeable enough about this topic. I'll do my best to do my homework and see if I can implement it in a reasonable time.

MrOneTwo replied:
I thought about this and I think I should not implement the PUSH method. That's because I have no way to test it in the foreseeable future. I wouldn't feel comfortable with merging this without testing it. I might be the one who implements that in the future but I can't do it right now.

MrOneTwo replied:
Since PUSH means just pushing data into the Software Management object, this should be supported.

jarlamsa reviewed (commented):
Thank you for the contribution. Here is an initial review before adapting our own solution to this object implementation. Might be that something arises when I start trying it out but I'll let you know.

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_pull_context.c:
```diff
@@ -344,6 +359,18 @@ static void firmware_transfer(struct k_work *work)
 	int ret;
 	char *server_addr;
 
```
fixed

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_pull_context.c:
```diff
@@ -0,0 +1,416 @@
+
+	/* Process incoming data */
+	payload_offset = response->hdr_len + response->opt_len;
+	coap_packet_get_payload(response, &payload_len);
```
Should be good now

DunderRoffe reviewed (changes requested):
Nice to see someone picking up my slack :)

I appreciate that you want to give me credit but as I am not the one doing this and or testing it I would appreciate if you would remove my name and sign-of-by line from the commits.

MrOneTwo commented:
@DunderRoffe Will do. Didn't forget, just didn't get to it yet :) Thanks for the comments, I'll start fixing that stuff soon.

MrOneTwo committed:
net: lwm2m: Add LWM2M object 9 software management

Adds support for LWM2M object 9 Software management.
This is implemented according to this release:
http://openmobilealliance.org/release/LWM2M_SWMGMT/V1_0_1-20200616-A/

Note that the XML is lacking some resources and for that reason those
resources are not included. This is a known problem by OMA and will be
fixed in a later releases.

This uses the lwm2m_pull_context to pull binaries in case
FIRMWARE_PULL_SUPPORT is enabled

Signed-off-by: Michal Ciesielski <michal.m.ciesielski@voiapp.io>

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,754 @@
+		return;
+	}
+
+	handle_event(instance, EVENT_DOWNLOAD_FAILED);
```
Could you elaborate why? For setting a correct initial state?

MrOneTwo commented:
@rlubos I think this is ready for you to take a look. I hope it's not a problem that I squashed the commits (if you have any pointers on how to make the review process easier then I'm all ears).

linusnielsen commented:
There is a small issue with this implementation of Object 9. There is no way to uninstall/delete the package after a reboot, because the state is always `UPD_STATE_INITIAL` at boot. There should probably be a way to set the initial state to `UPD_STATE_INSTALLED` in those cases where the package is already installed. Perhaps a callback? Or should the application use `lwm2m_engine_set_u8("/9/x/7", UPD_STATE_INSTALLED)` in these cases?

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,761 @@
+	error_code = lwm2m_pull_context_start_transfer(instance->package_uri, req, K_NO_WAIT);
+
+	if (error_code) {
+		return error_code;
```
Not sure here, this will result in an error response being sent to the server if some other download is in progress, wouldn't that be confusing for the server side (all in all URI write would succeed).

Shouldn't we rather have an extra delayed work item (per instance) and set it to retry download after some time in case of `EALREADY`?

MrOneTwo replied:
That's a good question... should we? Adding a delayed work would bump up the memory usage a bit for sure. Not sure about this one either.

rlubos replied:
I don't have a strong opinion here, I guess we coudl leave it as it is now (i. e. report an error to the server), and in cases it turns out problematic update in the future.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -458,30 +67,11 @@ int lwm2m_firmware_cancel_transfer(void)
-	firmware_ctx.fault_cb = socket_fault_cb;
-	firmware_retry = 0;
+	/* start file transfer work */
+	lwm2m_pull_context_start_transfer(package_uri, req, K_NO_WAIT);
```
I think this object should implement the same logic for handling start_transfer failures as SW mgmt.

MrOneTwo replied:
done

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,761 @@
+/* Package pull request should come with a verify_cb which needs to be stored for when package
+ * gets downloaded
+ */
+static int (*verify_package)(void);
```
This doesn't seem to be used at all (only set to NULL in one place), do we need it? I'd rather avoid introducing dead code, it might be confusing for folks analyzing the code.

linusnielsen replied:
It is called on line 582. However, I believe that callback should be part of `struct lwm2m_swmgmt_data`.

rlubos replied:
Well, I'd rather say the pointer is used, not called - it won't be called as it's only set to NULL in one place. And there's no API to change it value. But I agree with the latter, if it stays it should be part of the struct.

MrOneTwo replied:
I'm not sure about this. Why do we want to store a verify callback per object instance? What's the benefit there?
When I was porting object 9, the idea was to hold a verify callback per pull request. Mostly because this seems to do the job and is cheap memory wise. The API for that is the `struct requesting_object` - that's where one could set the `verify_cb`. Currently it's not really a user facing API but it conveys the idea of not storing a verification callback persistently, per object instance.

linusnielsen replied:
My thinking is that the verification might differ depending on the object requesting the pull. Each object might have a different way of checking the integrity, depending on the contents of the package.

rlubos reviewed (commented):
Thanks, I've left a few new comments, but overall I think we're pretty good here, nothing major. Good work with the rework, thanks!

MrOneTwo commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -264,6 +264,32 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 
+config LWM2M_SWMGMT_OBJ_SUPPORT
+	bool "Software management object support"
+	default y
```
fixed

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_pull_context.c:
```diff
@@ -0,0 +1,428 @@
+static char proxy_uri[LWM2M_PACKAGE_URI_LEN];
+#endif
+
+struct firmware_pull_context {
```
done

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_pull_context.c:
```diff
@@ -0,0 +1,428 @@
+#include "lwm2m_pull_context.h"
+#include "lwm2m_engine.h"
+
+K_SEM_DEFINE(lwm2m_pull_sem, 1, 1);
```
done

MrOneTwo commented on include/net/lwm2m.h:
```diff
@@ -385,14 +388,100 @@ void lwm2m_firmware_set_update_cb(lwm2m_engine_execute_cb_t cb);
+ * the corresponding object instance is called and the object is
+ * in the correct state.
+ *
+ * @param[in] obj_inst_id The instance number to set the callback for.
```
fixed. Please double check if the documentation is correct and understandable.

MrOneTwo replied:
Changed to `lwm2m_swmgmt_set_read_package_version_cb`

MrOneTwo commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,761 @@
+{
+	/*
+	 * TODO Implement handling for "ForUpdate".
+	 * This requires the Zephyr LWM2M to handle arguments to exec calls.
```
I'll delete this comment. I don't know what was the original author's intention here.

MrOneTwo replied:
deleted

rlubos commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -264,6 +264,32 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 
+config LWM2M_SWMGMT_OBJ_SUPPORT
+	bool "Software management object support"
+	default n
```
nit: We typically avoid `default n` since it's the implicit value.

MrOneTwo replied:
I'm a bit confused here. This change is in response to this comment https://github.com/zephyrproject-rtos/zephyr/pull/40552#discussion_r768668057
Is there a different way to turn this off without `default n`?

jarlamsa replied:
Removing that `default n` line will result as same behaviour.

rlubos replied:
Just delete the `default n` from the option, `default n` is the same as if there's no default at all, so since it's redundant we avoid it (tree-wise, it's not an lwm2m thing).

MrOneTwo replied:
ok, fixed

linusnielsen commented on subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:
```diff
@@ -0,0 +1,756 @@
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_obj_swmgmt
+#define LOG_LEVEL CONFIG_LWM2MM_LOG_LEVEL_DBG
```
Typo: Should be `CONFIG_LWM2M_LOG_LEVEL_DBG`

MrOneTwo replied:
And I do typing for a living 🤦 thanks
I changed this to `CONFIG_LWM2M_LOG_LEVEL`. It shouldn't be the `DBG` by default.

rlubos reviewed (approved):
No more comments from my side, thanks

rerickson1 reviewed (approved):
Looks great. Object 9 is on our list of requirements for our next project, looking forward to trying this out.

carlescufi commented:
@DunderRoffe @linusnielsen @jarlamsa could you please take another look and approve if appropriate?

carlescufi commented:
Please do not merge until @DunderRoffe has had a chance to take another look.

DunderRoffe reviewed (commented):
Commit messages and sign off looks good

Copyright notice from the firmware update object should probably still be added

MrOneTwo commented:
@DunderRoffe I looked through the code and the `lwm2m_obj_firmware_pull` code was mainly moved into the `lwm2m_pull_context`. Those two have the same copyright header. Are you saying I should add the `lwm2m_obj_firmware_pull` copyright header in the `lwm2m_obj_swmgmt.c`? Because that code seems to mostly be unique?
